### ISSUE TITLE
Using __typenames__ with a built-in type causes error

### ISSUE THREAD

**Issue by vpetrovykh** [edgedb/edgedb#2387]
- EdgeDB Version: 1.0-beta.2+dev.5540.g4142fc6c9.d20210319
- OS Version: Linux

Steps to Reproduce:

1. In REPL:
    ```
    edgedb> select [Object];
    ERROR: SchemaDefinitionError: cannot alter object type 'std::Object': module std is read-only
    ```

See `test_edgeql_select_array_common_type_02` in #2386

**Comment by msullivan** [comment]
The root cause here is that no view type is generated in this kind of case, and so we try to add the __tname fields to the actual real object type, which isn't allowed for base types. Breaks with tuples also.

**Comment by elprans** [comment]
Yeah, we have a check for this in `compile_query_subject`, but it only guards against directly selected objects.  One way to fix this would be to lower the check into `compile_path` and guard it with a context flag that turns it off when an explicit shape is present (setting it in `compile_Shape`, but perhaps there are other spots).

**Comment by msullivan** [comment]
Fixed by #3155
